File: PasswordHash.cs
using System;
using System.Text;
using System.Text.RegularExpressions;
using System.Security.Cryptography;

namespace HotFeet.Security.Cryptography {
	public class PasswordHash {
		// (...BitLength % 8 == 0)		
		static readonly int saltBitLength = 96;
		static readonly int hashBitLength = 512;

		static readonly string hashFormat = "$6${0}${1}";
		static readonly string base64Group = "(.{{{0}}})";
		static readonly string hashPattern = String.Format(
			hashFormat.Replace("$", @"\$"),
			String.Format(base64Group, toBase64Length(saltBitLength)),
			String.Format(base64Group, toBase64Length(hashBitLength))
		);
		static Regex hashRegex = new Regex(hashPattern);
		
		static readonly char zeroChar = Convert.ToBase64String(new byte[3] {0, 0, 0})[0];

		static int toBase64Length(int bitLength) {
			return (bitLength + 5) / 6;
		}
		
		static byte[] getBytes(string s) {
			return Encoding.UTF8.GetBytes(s);
		}

		HashAlgorithm hashAlg = HashAlgorithm.Create("SHA512");
		RandomNumberGenerator random = RandomNumberGenerator.Create();
		object hashAlgoLock = new object();
		
		static string ToFilledBase64(byte[] bytes) {
			int mod = bytes.Length % 3;
			if(mod == 0)
				return Convert.ToBase64String(bytes);

			int strLen = (bytes.Length * 8 + 5) / 6;
			int excess = 3 - mod;
			byte[] paddedBytes = new byte[bytes.Length + excess];
			Buffer.BlockCopy(bytes, 0, paddedBytes, 0, bytes.Length);
			for(int i = 0; i < excess; i++)
				paddedBytes[bytes.Length + i] = 0;
 
			return Convert.ToBase64String(paddedBytes).Substring(0, strLen);
		}

		static byte[] FromFilledBase64(string s) {
			int filledBits = s.Length * 6;
			if(filledBits % 8 == 0)
				return Convert.FromBase64String(s);

			//FIXME: don't hardcode the count
			s += new String(zeroChar, 2);
			byte[] bytesExcessive = Convert.FromBase64String(s);
			byte[] bytes = new byte[filledBits >> 3];
			Array.Copy(bytesExcessive, bytes, bytes.Length);

			return bytes;
		}

		public string Hash(string password) {
			byte[] salt = new byte[saltBitLength >> 3];
			random.GetBytes(salt);
				
			byte[] pass = getBytes(password);
			byte[] hash = Hash(pass, salt);
				
			return String.Format(
				hashFormat,
				ToFilledBase64(salt),
				ToFilledBase64(hash)
			);
		}
		
		byte[] Hash(byte[] password, byte[] salt) {
			byte[] plain = new byte[password.Length + 2 * salt.Length];
			Buffer.BlockCopy(salt, 0, plain, 0, salt.Length);
			Buffer.BlockCopy(password, 0, plain, salt.Length, password.Length);
			Buffer.BlockCopy(salt, 0, plain, salt.Length + password.Length, salt.Length);

			//HashAlgorithm is stateful 
			byte[] hash;
			lock(hashAlgoLock) {
				hashAlg.TransformFinalBlock(plain, 0, plain.Length);
				hash = hashAlg.Hash;

				hashAlg.Clear();
			}
			return hash;
		}
		
		public bool Verify(string password, string hash) {
			Match m = hashRegex.Match(hash);
			if(!m.Success)
				throw new ArgumentException("codedHash");

			byte[] storedSalt = FromFilledBase64(m.Groups[1].Value);
			byte[] storedHash = FromFilledBase64(m.Groups[2].Value);

			byte[] freshHash = Hash(getBytes(password), storedSalt);
			for(int i = 0; i < freshHash.Length; i++) {
				if(freshHash[i] != storedHash[i])
					return false;
			}
			
			return true;
		}
	}
	
	static class Driver {
		static void Main(string[] args) {
			var ph = new PasswordHash();
			string hash = ph.Hash(args[0]);
			Console.WriteLine(hash);
			Console.WriteLine(ph.Verify(args[0], hash));
		}
	}
}

File: PasswordHash.fs
//
// PasswordHash
//
// Authors:
//    Juraj Skripsky <js@hotfeet.ch>
//
// Copyright 2011 HotFeet GmbH (http://www.hotfeet.ch)
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

open System
open System.Text
open System.Security.Cryptography
open HotFeet.Text
open HotFeet.FSharp

namespace HotFeet.Security.Cryptography
	type public PasswordHash () =
		(* input/ouput character encoding *)
		let enc = Encoding.UTF8
		let toBytes (s: string) = enc.GetBytes (s)
		let toString (bs: byte[]) = enc.GetString (bs)
		 
		(* Crypto primitives *)
		let hashAlgoName = "SHA512"		
		let hashBitLen = 512 //(len % 8 = 0)
		let saltBitLen = 96 //(len % 8 = 0)

		(* Hash Algorithm *)	
		let (hashAlgo, hashAlgoLock) = (HashAlgorithm.Create (hashAlgoName), new obj())
		let hashBytes bs =
			let calcHash =
				let (_, hash) = (hashAlgo.TransformFinalBlock (bs, 0, bs.Length), hashAlgo.Hash)
				hashAlgo.Clear()
				hash
			lock hashAlgoLock (fun _ -> calcHash)
			 
		(* Random Number Generator *)	
		let (randGen, randGenLock) = (RandomNumberGenerator.Create (), new obj())
		let random (bs : byte[]) =
			lock randGenLock (fun _ -> randGen.GetBytes (bs))
			bs
	
		let randomBytes len =
			let mutable (bs : byte[]) = Array.zeroCreate len
			random bs
		
		let saltPassword password salt =
			let (+) x y = Array.append x y
			salt + password + salt
	
		(* crypted password format *)
		let format = sprintf "$6$%s$%s"
		let cryptedPasswordRegex =
			let base64Group bitLen = sprintf "(.{%d})" (bitLen |> Base64.getStringLen)
			let pattern = format (base64Group saltBitLen) (base64Group hashBitLen)
			regex (pattern.Replace ("$", @"\$")) 
	
		(* input / output *)
		let compose (salt, hash) = format (salt |> Base64.toString) (hash |> Base64.toString) 
		let decompose s =
			match s with
			| Match cryptedPasswordRegex [salt; hash] -> (salt |> Base64.ofString, hash |> Base64.ofString)
			| _ -> failwith "Invalid hash format."
	
		(* main methods *)
		let crypt password =
			let salt = randomBytes (saltBitLen >>> 3)
			(salt, password salt ||> saltPassword > hashBytes)
		
		let verify password (salt, hash) = (password salt ||> saltPassword > hashBytes) = hash
	
		interface IPasswordHash with
			member x.Crypt password = password |> toBytes |> crypt |> compose
			member x.Verify (password, cryptedPassword) = (password >> toBytes, cryptedPassword |> decompose) ||> verify
